### ISSUE TITLE
Plan autogeneration of documentation and its hosting

### ISSUE THREAD

**Issue by asoplata** [DynaSim/DynaSim#18]
1. I (or someone else) need to evaluate if [readthedocs](https://readthedocs.org/) satisfies our User Documentation and/or Programmer Documentation requirements. It almost certainly does, but if, e.g., it doesn't allow picture posting, then we may want to host our User Documentation somewhere else that does, like this repo's GitHub wiki.
2. I need to implement a super simple workflow for autogeneration of documentation that's compatible with the site. This is tied to #17 so we'll probably want to use Doxygen or some trivial way to produce whatever the output files are, and then upload them to readthedocs, however that's done...

**Comment by asoplata** [comment]
Given that [readthedocs](readthedocs.org)
    1. looks real good
    2. supports pictures
    3. relies on the popular [Sphinx](http://www.sphinx-doc.org/en/stable/) documentation generator to actually power it / as the standard. I'm not even sure IF it supports other documentation generation methods.
    4. is a popular source for documentation for GitHub repos

and, given that
    1. Sphinx AFAIK is popular in the Python community
    2. There already exists a [Sphinx "filter" for MATLAB](https://www.mathworks.com/matlabcentral/linkexchange/links/3703-matlab-sphinx-documentation),

I believe 
    1. we should go with readthedocs
    2. we should use the Sphinx doc gen system.

If anyone has problems with this decision, this is the place to air them.

### TODO:
1. Write "Expectations" in the Style section of the Programmer Documentation to include that anyone who changes the behavior of the code in such a way that it affects the documentation is responsible for also updating the documentation, at least some point before it is merged into an important branch like master. I don't know if we should be adamant that documentation-changing should be done in the same commit as the actual code change -- probably.
2. Write a somewhat-cross-platform bash script for trivial updating of the docs, after said change. This possibly could be done as a git hook, if it can be done cross-platform.

**Comment by asoplata** [comment]
Sigh.

**tl;dr: Sphinx/readthedocs.org is too difficult, but also not functional enough, to justify using it for our autodocumentation. Instead, we should use [m2html](http://www.artefact.tk/software/matlab/m2html/) and just throw its HTML/etc files onto a [GitHub Project Pages website](https://help.github.com/articles/user-organization-and-project-pages/) for DynaSim directly.**

Long version:

After MUCH playing around with Sphinx and the MATLAB Sphinx extension in [this Issue's branch](https://github.com/DynaSim/DynaSim/tree/iss0017_docu_std), and discussion with @erikthered12 over more general approaches to documentation in DynaSim, I've changed my mind and strongly believe we should NOT use Sphinx and its extension for DynaSim. Problems I've run into:

1. We'd have to reformat all existing documentation (not that difficult with a good text editor).
2. The Sphinx extension mentioned previously only works with Python 2.7, and thus anyone who wants to autogenerate the documentation would have to install the right Python version and its libraries (difficult depending on platform, and very not newbie-friendly).
3. Sphinx uses [reStructuredText](http://docutils.sourceforge.net/rst.html) which, while very well-defined, I find to be NOT intuitive AT ALL, and unnecessarily complicated for our project. It's extremely whitespace/syntax-sensitive, and it would not be clear to individuals how to expand documentation without learning reST.
4. Furthermore, the Sphinx extension itself is "beta" software that can only use a fraction of Sphinx's power, and is also no longer under active development. As far as I've been able to test, it is pretty much only useful for pretty-printing individual function documentation in our case and that singular thing ONLY.

Thus, in taking a second look at using Sphinx/Readthedocs.org for our project, I think there are far easier and better solutions to be had:

#### Documentation autogeneration possibilities:

1. [MATLAB's native `publish`](https://www.mathworks.com/help/matlab/ref/publish.html) functionality. Short version: **I don't think we should use this**, `publish` really doesn't give us very much at all, MATLAB doesn't seem to care about that, and we would still have to organize the overall structure of any resulting documentation ourselves. Long version: I've spent the last couple hours trying to evaluate the full functionality of function help self-documentation in core MATLAB, and it's quite limited. While MATLAB can `publish` to many different formats (useful if we want to build documentation to PDF), there is very little integration/functionality one can implement via help documentation in their own functions in MATLAB (e.g. https://www.mathworks.com/help/matlab/matlab_prog/marking-up-matlab-comments-for-publishing.html). As far as I can tell, the native help documentation you can generate (insofar as "functionality") is simply: 1. You can make a multi-line comment after the function definition that gets returned when you run `help <function>`, 2. if you add "nearby" functions etc. to a help-comment line starting with "See also", upon `help <function>` those functions will be hyperlinked, and 3. If you add two comment symbols to a section (`%%`), when you `publish` a hyperlink to that section will be generated at the top of the file. Example output is shown here ![screenshot from 2017-01-31 20-24-53](https://cloud.githubusercontent.com/assets/661173/22492243/d4d632d8-e7f6-11e6-8e98-deaf46b59ade.png). There are also MATLAB bugs apparently with publishing code segmented by the double-comment-symbol, which I won't get into (it seems to think of them as completely independent sections?). I've also desperately tried to hunt for officially-recommended MATLAB documentation standards (e.g. INPUTS/OUTPUTS), and I've hunted FAR and WIDE, even in the Internet Archive, but it seems MATLAB has severely cut down their documentation (and possibly support) of self-documentation. Additionally, `publish` seems to be focused towards individual scripts and files, rather than toolboxes or projects. There is a little toolbox-assisting documentation functionality, but it's so small that I can barely remember what is it (I do remember that I couldn't get the "search database" functionality to work). Currently, using the built-in tools for MATLAB autodocumentation seems the worse choice compared to the next option...
2. !!! The third-party [m2html](http://www.artefact.tk/software/matlab/m2html/) Example output is shown here ![screenshot from 2017-01-31 20-25-06](https://cloud.githubusercontent.com/assets/661173/22492474/021a0548-e7f8-11e6-8667-8af1db8b8a00.png). This works out-of-the-box, albeit slowly: Not only does this
    1. Nicely format individual function documentation into individual HTML files, but it also
    2. Only needs to be run once to document the entire codebase (as opposed to `publish` where we'd have to write some code to do that and organize it all),
    3. Thereby organizing all the HTML/etc. so we can trivially upload it all to a website in one go
    4. Automatically cross-references each function (producing hyperlinks to every other function called by any one function) AND separately identifies subfunctions!!!
    5. If you have Graphviz installed, it will also automatically generate a function dependency graph like we've been doing ![graph](https://cloud.githubusercontent.com/assets/661173/22492680/16d064c2-e7f9-11e6-9370-2961451d684f.png)
3. Using [Doxygen somehow](https://www.mathworks.com/matlabcentral/fileexchange/25925-using-doxygen-with-matlab), which is similar to Sphinx. This would also require us to commit to a fundamentally different format for our documentation, we'd probably have to "build" the documentation using a non-MATLAB language, I have no idea what the complexity level is of Doxygen formatting, and I also don't know how limited/functional the MATLAB->Doxygen tools are. In short, this likely would be close to the failed Sphinx method in complexity. Furthermore, theoretically we could use this on Readthedocs.org [(see here)](http://breathe.readthedocs.io/en/latest/readthedocs.html) but that seems to be experimental, and I have no idea if it would work with MATLAB docs.

For many reasons, I think m2html is the FAR, FAR superior choice, especially since it's trivial to run, works out of the box, and looks great out of the box. `publish` is pretty limited compared to m2html, and Doxygen in all likelihood will be just as complex as the Sphinx method I've already tried (though it probably won't be MORE complicated). If we want to take m2html's HTML/etc. and host it on a website, all we need to do is copy it to the hosting server and we're done!

I will mention the single caveat I have about m2html: **it's slow**, and while I've used it in years past and it was NOT slow, on my desktop at CompNet, it took about **7.5 minutes** to generate all the DynaSim documentation from scratch -- that's all HTML/etc. pages, and calculating the function graph, everything (should we write a batch script for this? lol jk). This is relatively old software, and I saw a warning about particular functions it uses which are soon to be deprecated, so doing our own custom updates to the software may speed it up considerably. This isn't that much of an issue, since the documentation doesn't need to get built super frequently, but we may want to see if we can adapt/speed up m2html ourselves. Even with this in mind though, m2html is CLEARLY the best current option for autogenerating the documentation IMHO.

#### Hosting

On that note, formerly we were set on Readthedocs.org since it's so commonly used and we thought Sphinx would be easy to convert to. Instead, since m2html gives us a folder with a nice little index.html and all its function HTML files already hyperlinked, I think that (as @erikthered12 suggested) we should look at throwing the m2html docs up on a [GitHub Project Pages site](https://help.github.com/articles/user-organization-and-project-pages/) since it should really be as easy as copying them into a subfolder. This will definitely need to be tested, however.

#### Conclusion

See the **tl;dr** above.

**Comment by kupiqu** [comment]
Cool!

**Comment by erikthered12** [comment]
@asoplata, how did you figure out the nargin check thing?

**Comment by asoplata** [comment]
When you try to run m2html in MATLAB 2016B, it spits out hundreds of errors
that NARGCHK will be deprecated. I emailed the developer and they say
they're putting m2html on github soon. I suspect that it's the error and
warning raising that was causing the slowdown.

**Comment by mnarayan** [comment]
If you like m2html then you might be interested in python based [matlabdoc](https://github.com/daviskirk/matlabdoc) which is modeled after m2html.
